# Лабораторная работа 4. Обучение с учителем. Задача классификации.

# Задание 1.1

Выполнено в ноутбуке Lab4_Task1.ipynb

# Исследование данных

In [40]:
import pandas as pd

In [41]:
df = pd.read_csv('../../../Datasets/Lab4_DataSet4_Preprocessed.csv')

In [42]:
df.head()

id  Airline_0  Airline_1  Airline_2  Airline_3  Airline_4  Flight   
0  400740          0          0          0          0          1    3036  \
1  194808          0          0          0          1          0     315   
2  526903          0          0          0          1          1    6673   
3  477937          0          0          1          0          0    3668   
4   48180          0          0          1          0          1     623   

   AirportFrom_0  AirportFrom_1  AirportFrom_2  ...  AirportTo_3  AirportTo_4   
0              0              0              0  ...            0            0  \
1              0              0              0  ...            0            0   
2              0              0              0  ...            0            0   
3              0              0              0  ...            0            0   
4              0              0              0  ...            0            0   

   AirportTo_5  AirportTo_6  AirportTo_7  AirportTo_8  DayOfWeek  Time   
0            0            0            0            1          4  1195  \
1            0            0            1            0          7   707   
2            0            0            1            1          5   499   
3            0            0            1            0          2   810   
4            0            1            0            0          5   985   

   Length  Delay  
0     131      0  
1     145      0  
2      42      1  
3      70      1  
4     143      0  

[5 rows x 29 columns]

In [43]:
df = df.drop('id', axis=1)

In [44]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 399941 entries, 0 to 399940
Data columns (total 28 columns):
 #   Column         Non-Null Count   Dtype
---  ------         --------------   -----
 0   Airline_0      399941 non-null  int64
 1   Airline_1      399941 non-null  int64
 2   Airline_2      399941 non-null  int64
 3   Airline_3      399941 non-null  int64
 4   Airline_4      399941 non-null  int64
 5   Flight         399941 non-null  int64
 6   AirportFrom_0  399941 non-null  int64
 7   AirportFrom_1  399941 non-null  int64
 8   AirportFrom_2  399941 non-null  int64
 9   AirportFrom_3  399941 non-null  int64
 10  AirportFrom_4  399941 non-null  int64
 11  AirportFrom_5  399941 non-null  int64
 12  AirportFrom_6  399941 non-null  int64
 13  AirportFrom_7  399941 non-null  int64
 14  AirportFrom_8  399941 non-null  int64
 15  AirportTo_0    399941 non-null  int64
 16  AirportTo_1    399941 non-null  int64
 17  AirportTo_2    399941 non-null  int64
 18  AirportTo_3    399941 no

In [45]:
df = df.drop_duplicates().reset_index()

# Задание 2.1

Выделите целевой признак и предикторы, разбейте данные на обучающую и тестовую выборку.

## Целевой признак и предикторы

In [46]:
X = df.drop('Delay',axis=1)
Y = df['Delay']

In [47]:
X.head()

index  Airline_0  Airline_1  Airline_2  Airline_3  Airline_4  Flight   
0      0          0          0          0          0          1    3036  \
1      1          0          0          0          1          0     315   
2      2          0          0          0          1          1    6673   
3      3          0          0          1          0          0    3668   
4      4          0          0          1          0          1     623   

   AirportFrom_0  AirportFrom_1  AirportFrom_2  ...  AirportTo_2  AirportTo_3   
0              0              0              0  ...            0            0  \
1              0              0              0  ...            0            0   
2              0              0              0  ...            0            0   
3              0              0              0  ...            0            0   
4              0              0              0  ...            0            0   

   AirportTo_4  AirportTo_5  AirportTo_6  AirportTo_7  AirportTo_8  DayOfWeek   
0            0            0            0            0            1          4  \
1            0            0            0            1            0          7   
2            0            0            0            1            1          5   
3            0            0            0            1            0          2   
4            0            0            1            0            0          5   

   Time  Length  
0  1195     131  
1   707     145  
2   499      42  
3   810      70  
4   985     143  

[5 rows x 28 columns]

In [48]:
Y.head()

0    0
1    0
2    1
3    1
4    0
Name: Delay, dtype: int64

In [49]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(X,Y, test_size=0.1,random_state=42)

x_train.shape, x_test.shape, y_train.shape, y_test.shape

((229500, 28), (25500, 28), (229500,), (25500,))

# Задание 2.2

- Масштабируйте числовые признаки.

In [50]:
from sklearn.preprocessing import MinMaxScaler
import numpy as np

scaler = MinMaxScaler(feature_range=(0,1))
X['Flight'] = scaler.fit_transform(np.reshape(X['Flight'],(-1,1)))
df['Flight'] = scaler.transform(np.reshape(df['Flight'],(-1,1)))

x_train['Flight'] = scaler.transform(np.reshape(x_train['Flight'],(-1,1)))
x_test['Flight'] = scaler.transform(np.reshape(x_test['Flight'],(-1,1)))

# Задание 2.3

- В случае дисбаланса проведите балансировку.

In [51]:
x_train.head()

index  Airline_0  Airline_1  Airline_2  Airline_3  Airline_4   
188572  253792          0          0          0          1          1  \
142495  175674          0          0          0          0          1   
77726    86490          0          1          0          0          1   
234930  350310          0          0          0          1          0   
45243    48007          0          0          1          0          0   

          Flight  AirportFrom_0  AirportFrom_1  AirportFrom_2  ...   
188572  0.816436              0              0              0  ...  \
142495  0.325781              0              0              0  ...   
77726   0.149386              0              0              0  ...   
234930  0.001152              0              0              0  ...   
45243   0.131720              0              0              0  ...   

        AirportTo_2  AirportTo_3  AirportTo_4  AirportTo_5  AirportTo_6   
188572            0            1            1            1            1  \
142495            0            0            1            1            0   
77726             1            0            0            0            0   
234930            0            0            1            0            0   
45243             1            0            0            0            0   

        AirportTo_7  AirportTo_8  DayOfWeek  Time  Length  
188572            1            0          3   467      54  
142495            1            0          4   370     126  
77726             0            1          4  1360     285  
234930            0            0          3   645     344  
45243             0            1          4  1145      55  

[5 rows x 28 columns]

In [52]:
Y[Y == 0].shape, Y[Y == 1].shape

((168109,), (86891,))

In [53]:
y_test[y_test == 0].shape, y_test[y_test == 1].shape

((16823,), (8677,))

In [54]:
from sklearn.utils import resample

neg_class = df[df['Delay'] == 0]
pos_class = df[df['Delay'] == 1]

neg_downsampled_class = resample(neg_class,n_samples=len(pos_class),random_state=42)
downsampled_df = pd.concat((neg_downsampled_class,pos_class),axis=0)

downsampled_df.head()

index  Airline_0  Airline_1  Airline_2  Airline_3  Airline_4   
174996  229288          0          1          1          0          1  \
216142  308310          0          0          1          0          0   
191202  258754          0          0          0          1          1   
146625  182023          0          1          0          0          1   
171653  223353          0          1          1          1          0   

          Flight  AirportFrom_0  AirportFrom_1  AirportFrom_2  ...   
174996  0.201101              0              0              0  ...  \
216142  0.373016              0              0              0  ...   
191202  0.837302              0              0              0  ...   
146625  0.166923              0              0              0  ...   
171653  0.688940              0              1              1  ...   

        AirportTo_3  AirportTo_4  AirportTo_5  AirportTo_6  AirportTo_7   
174996            0            1            1            1            1  \
216142            1            1            0            0            1   
191202            0            0            1            1            0   
146625            0            0            0            1            1   
171653            0            0            1            1            1   

        AirportTo_8  DayOfWeek  Time  Length  Delay  
174996            0          3   735      75      0  
216142            1          5   635      90      0  
191202            1          5   765      57      0  
146625            0          6   600     163      0  
171653            1          4   430     160      0  

[5 rows x 29 columns]

In [55]:
X = downsampled_df.drop('Delay', axis=1)
Y = downsampled_df['Delay']

x_train, x_test, y_train, y_test = train_test_split(X,Y,random_state=42,test_size=0.1)

In [56]:
y_train[y_train == 0].shape, y_train[y_train == 1].shape

((78216,), (78187,))

# Задание 2.4

- Решите задачу классификации на ваших данных с использованием рассмотренных моделей Sklearn: k-NN, Naive Bayes, Logistic Regression, SVM. Не забудьте подобрать гиперпараметры.

- Вычислите значения метрик Accuracy, Precision, Recall, F1, ROC AUC score.

In [57]:
from sklearn.metrics import confusion_matrix,classification_report,roc_auc_score,accuracy_score

## k-NN

### 5 соседей

In [58]:
from sklearn.neighbors import KNeighborsClassifier

knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(x_train,y_train)
predict = knn.predict(x_test)

print(confusion_matrix(y_test,predict))
print('\n',classification_report(y_test,predict))
print('roc_auc_score:',roc_auc_score(y_test,predict))

[[5121 3554]
 [3880 4824]]

               precision    recall  f1-score   support

           0       0.57      0.59      0.58      8675
           1       0.58      0.55      0.56      8704

    accuracy                           0.57     17379
   macro avg       0.57      0.57      0.57     17379
weighted avg       0.57      0.57      0.57     17379

roc_auc_score: 0.5722724720291574


### Подбор гиперпараметров к KNN

In [66]:
from sklearn.model_selection import RandomizedSearchCV

neighbors = np.arange(3,16,2)
params  = dict(n_neighbors=neighbors)
params

{'n_neighbors': array([ 3,  5,  7,  9, 11, 13, 15])}

In [67]:
knn = KNeighborsClassifier()
randomized_search = RandomizedSearchCV(knn,params,scoring='accuracy',n_jobs=-1,refit=True)
randomized_search.fit(x_train,y_train)

c:\Python\Lib\site-packages\sklearn\model_selection\_search.py:307: UserWarning: The total space of parameters 7 is smaller than n_iter=10. Running 7 iterations. For exhaustive searches, use GridSearchCV.
  warnings.warn(


RandomizedSearchCV(estimator=KNeighborsClassifier(), n_jobs=-1,
                   param_distributions={'n_neighbors': array([ 3,  5,  7,  9, 11, 13, 15])},
                   scoring='accuracy')

In [69]:
randomized_search.best_params_

{'n_neighbors': 3}

In [71]:
predict = randomized_search.predict(x_test)

In [72]:
print(confusion_matrix(y_test,predict))
print('\n',classification_report(y_test,predict))
print('roc_auc_score:',roc_auc_score(y_test,predict))

[[5233 3442]
 [3936 4768]]

               precision    recall  f1-score   support

           0       0.57      0.60      0.59      8675
           1       0.58      0.55      0.56      8704

    accuracy                           0.58     17379
   macro avg       0.58      0.58      0.58     17379
weighted avg       0.58      0.58      0.58     17379

roc_auc_score: 0.5755108916765553


## Naive Bayes метод

In [73]:
from sklearn.naive_bayes import GaussianNB

gauss = GaussianNB()
gauss.fit(x_train,y_train)
predict = gauss.predict(x_test)

print(confusion_matrix(y_test,predict))
print('\n',classification_report(y_test,predict))
print('roc_auc_score:',roc_auc_score(y_test,predict))

[[5156 3519]
 [3633 5071]]

               precision    recall  f1-score   support

           0       0.59      0.59      0.59      8675
           1       0.59      0.58      0.59      8704

    accuracy                           0.59     17379
   macro avg       0.59      0.59      0.59     17379
weighted avg       0.59      0.59      0.59     17379

roc_auc_score: 0.5884786417719104


## Logistic Regression метод

In [75]:
from sklearn.linear_model import LogisticRegression

lr = LogisticRegression()
lr.fit(x_train,y_train)
predict = lr.predict(x_test)

print(confusion_matrix(y_test,predict))
print('\n',classification_report(y_test,predict))
print('roc_auc_score:',roc_auc_score(y_test,predict))

[[3710 4965]
 [2817 5887]]

               precision    recall  f1-score   support

           0       0.57      0.43      0.49      8675
           1       0.54      0.68      0.60      8704

    accuracy                           0.55     17379
   macro avg       0.56      0.55      0.55     17379
weighted avg       0.56      0.55      0.55     17379

roc_auc_score: 0.5520107022906424


### Подбор гиперпараметров к Logistic Regression

In [78]:
lr = LogisticRegression()
C = np.logspace(-4, 4, 50)
params = dict(
    C=C,
)

randomized_search = RandomizedSearchCV(lr,params,n_jobs=-1)
randomized_search.fit(x_train,y_train)

RandomizedSearchCV(estimator=LogisticRegression(), n_jobs=-1,
                   param_distributions={'C': array([1.00000000e-04, 1.45634848e-04, 2.12095089e-04, 3.08884360e-04,
       4.49843267e-04, 6.55128557e-04, 9.54095476e-04, 1.38949549e-03,
       2.02358965e-03, 2.94705170e-03, 4.29193426e-03, 6.25055193e-03,
       9.10298178e-03, 1.32571137e-02, 1.93069773e-02, 2.81176870e-02,
       4.09491506e-02, 5.9...
       8.28642773e-01, 1.20679264e+00, 1.75751062e+00, 2.55954792e+00,
       3.72759372e+00, 5.42867544e+00, 7.90604321e+00, 1.15139540e+01,
       1.67683294e+01, 2.44205309e+01, 3.55648031e+01, 5.17947468e+01,
       7.54312006e+01, 1.09854114e+02, 1.59985872e+02, 2.32995181e+02,
       3.39322177e+02, 4.94171336e+02, 7.19685673e+02, 1.04811313e+03,
       1.52641797e+03, 2.22299648e+03, 3.23745754e+03, 4.71486636e+03,
       6.86648845e+03, 1.00000000e+04])})

In [79]:
randomized_search.best_params_

{'C': 3.727593720314938}

In [82]:
predict = randomized_search.predict(x_test)

print(confusion_matrix(y_test,predict))
print('\n',classification_report(y_test,predict))
print('roc_auc_score:',roc_auc_score(y_test,predict))

[[3710 4965]
 [2817 5887]]

               precision    recall  f1-score   support

           0       0.57      0.43      0.49      8675
           1       0.54      0.68      0.60      8704

    accuracy                           0.55     17379
   macro avg       0.56      0.55      0.55     17379
weighted avg       0.56      0.55      0.55     17379

roc_auc_score: 0.5520107022906424


## SVM 

In [83]:
from sklearn.svm import LinearSVC
from sklearn.linear_model import SGDClassifier

### Линейная svm

In [84]:
linear_svc = LinearSVC()
linear_svc.fit(x_train,y_train)
predict = linear_svc.predict(x_test)

print(confusion_matrix(y_test,predict))
print('\n',classification_report(y_test,predict))
print('roc_auc_score:',roc_auc_score(y_test,predict))

c:\Python\Lib\site-packages\sklearn\svm\_classes.py:32: FutureWarning: The default value of `dual` will change from `True` to `'auto'` in 1.5. Set the value of `dual` explicitly to suppress the warning.
  warnings.warn(


[[ 738 7937]
 [ 318 8386]]

               precision    recall  f1-score   support

           0       0.70      0.09      0.15      8675
           1       0.51      0.96      0.67      8704

    accuracy                           0.53     17379
   macro avg       0.61      0.52      0.41     17379
weighted avg       0.61      0.53      0.41     17379

roc_auc_score: 0.524268559819461


c:\Python\Lib\site-packages\sklearn\svm\_base.py:1250: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(


### SVM с методом градиентного спуска

In [85]:
sgd = SGDClassifier()
sgd.fit(x_train,y_train)
predict = sgd.predict(x_test)

print(confusion_matrix(y_test,predict))
print('\n',classification_report(y_test,predict))
print('roc_auc_score:',roc_auc_score(y_test,predict))

[[8675    0]
 [8704    0]]

               precision    recall  f1-score   support

           0       0.50      1.00      0.67      8675
           1       0.00      0.00      0.00      8704

    accuracy                           0.50     17379
   macro avg       0.25      0.50      0.33     17379
weighted avg       0.25      0.50      0.33     17379

roc_auc_score: 0.5


c:\Python\Lib\site-packages\sklearn\metrics\_classification.py:1469: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
c:\Python\Lib\site-packages\sklearn\metrics\_classification.py:1469: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
c:\Python\Lib\site-packages\sklearn\metrics\_classification.py:1469: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


### Подбор гиперпараметров к SVM

#### LinearSVC

In [87]:
linear_svc = LinearSVC()
params = dict(
    penalty=['l1','l2'],
    C=np.array([1,10,20,40,80,100]),
)

randomized_search = RandomizedSearchCV(linear_svc,params,n_jobs=-1)
randomized_search.fit(x_train,y_train)

c:\Python\Lib\site-packages\sklearn\model_selection\_validation.py:425: FitFailedWarning: 
30 fits failed out of a total of 50.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
30 fits failed with the following error:
Traceback (most recent call last):
  File "c:\Python\Lib\site-packages\sklearn\model_selection\_validation.py", line 729, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "c:\Python\Lib\site-packages\sklearn\base.py", line 1152, in wrapper
    return fit_method(estimator, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "c:\Python\Lib\site-packages\sklearn\svm\_classes.py", line 326, in fit
    self.coef_, self.intercept_, n_iter_ = _fit_liblinear(
                          

RandomizedSearchCV(estimator=LinearSVC(), n_jobs=-1,
                   param_distributions={'C': array([  1,  10,  20,  40,  80, 100]),
                                        'penalty': ['l1', 'l2']})

In [88]:
randomized_search.best_params_

{'penalty': 'l2', 'C': 20}

In [89]:
predict = randomized_search.predict(x_test)

print(confusion_matrix(y_test,predict))
print('\n',classification_report(y_test,predict))
print('roc_auc_score:',roc_auc_score(y_test,predict))

[[ 453 8222]
 [ 277 8427]]

               precision    recall  f1-score   support

           0       0.62      0.05      0.10      8675
           1       0.51      0.97      0.66      8704

    accuracy                           0.51     17379
   macro avg       0.56      0.51      0.38     17379
weighted avg       0.56      0.51      0.38     17379

roc_auc_score: 0.5101972858217495


#### SGDClassifier

In [90]:
sgd = SGDClassifier()
params = dict(
    eta0=[0.00001, 0.0001, 0.001, 0.01, 0.1, 1],
    learning_rate=['constant'],
    alpha=[0.00001, 0.0001, 0.001, 0.01, 0.1, 1]
)

randomized_search = RandomizedSearchCV(sgd,params,n_jobs=-1)
randomized_search.fit(x_train,y_train)

RandomizedSearchCV(estimator=SGDClassifier(), n_jobs=-1,
                   param_distributions={'alpha': [1e-05, 0.0001, 0.001, 0.01,
                                                  0.1, 1],
                                        'eta0': [1e-05, 0.0001, 0.001, 0.01,
                                                 0.1, 1],
                                        'learning_rate': ['constant']})

In [91]:
randomized_search.best_params_

{'learning_rate': 'constant', 'eta0': 1e-05, 'alpha': 0.001}

In [92]:
predict = randomized_search.predict(x_test)
print(confusion_matrix(y_test,predict))
print('\n',classification_report(y_test,predict))
print('roc_auc_score:',roc_auc_score(y_test,predict))

[[ 180 8495]
 [ 129 8575]]

               precision    recall  f1-score   support

           0       0.58      0.02      0.04      8675
           1       0.50      0.99      0.67      8704

    accuracy                           0.50     17379
   macro avg       0.54      0.50      0.35     17379
weighted avg       0.54      0.50      0.35     17379

roc_auc_score: 0.5029642537400407


# Задание 2.5

- Построить итоговую модель классификаторов (см. Lab4_Ex5 ClassificationModel.ipynb).

In [93]:
classifiers = {
    'KNN метод': KNeighborsClassifier(),
    'Naive Bayes метод': GaussianNB(),
    'Logistic Regression метод': LogisticRegression(),
    'LinearSVC': LinearSVC(),
    'SGDClassifier': SGDClassifier()
}

n_classifiers = len(classifiers)

for index, (name, classifier) in enumerate(classifiers.items()):
    classifier.fit(x_train, np.ravel(y_train))

    y_pred = classifier.predict(x_test)
    accuracy = accuracy_score(y_test, y_pred)
    print("Accuracy (train) for %s: %0.1f%% " % (name, accuracy * 100))
    print(classification_report(y_test,y_pred))

Accuracy (train) for KNN метод: 57.2% 
              precision    recall  f1-score   support

           0       0.57      0.59      0.58      8675
           1       0.58      0.55      0.56      8704

    accuracy                           0.57     17379
   macro avg       0.57      0.57      0.57     17379
weighted avg       0.57      0.57      0.57     17379

Accuracy (train) for Naive Bayes метод: 58.8% 
              precision    recall  f1-score   support

           0       0.59      0.59      0.59      8675
           1       0.59      0.58      0.59      8704

    accuracy                           0.59     17379
   macro avg       0.59      0.59      0.59     17379
weighted avg       0.59      0.59      0.59     17379

Accuracy (train) for Logistic Regression метод: 55.2% 
              precision    recall  f1-score   support

           0       0.57      0.43      0.49      8675
           1       0.54      0.68      0.60      8704

    accuracy                           0.

c:\Python\Lib\site-packages\sklearn\svm\_classes.py:32: FutureWarning: The default value of `dual` will change from `True` to `'auto'` in 1.5. Set the value of `dual` explicitly to suppress the warning.
  warnings.warn(
c:\Python\Lib\site-packages\sklearn\svm\_base.py:1250: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(


Accuracy (train) for LinearSVC: 52.9% 
              precision    recall  f1-score   support

           0       0.61      0.16      0.26      8675
           1       0.52      0.90      0.66      8704

    accuracy                           0.53     17379
   macro avg       0.56      0.53      0.46     17379
weighted avg       0.56      0.53      0.46     17379

Accuracy (train) for SGDClassifier: 49.9% 
              precision    recall  f1-score   support

           0       0.50      1.00      0.67      8675
           1       0.00      0.00      0.00      8704

    accuracy                           0.50     17379
   macro avg       0.25      0.50      0.33     17379
weighted avg       0.25      0.50      0.33     17379



c:\Python\Lib\site-packages\sklearn\metrics\_classification.py:1469: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
c:\Python\Lib\site-packages\sklearn\metrics\_classification.py:1469: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
c:\Python\Lib\site-packages\sklearn\metrics\_classification.py:1469: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


# Задание 3.1

- Самостоятельно реализуйте метрики Accuracy, Precision, Recall, F1

In [94]:
class MyMetrics:
    def __init__(self,y_true,y_pred) -> None:
        self.y_true = np.reshape(y_true,(-1,1))
        self.y_pred = np.reshape(y_pred,(-1,1))
        self.confusion_matrix_ = np.zeros((2,2)).astype(int)

        vector = np.concatenate((self.y_true,self.y_pred),axis=1)
        for x,y in vector:
            self.confusion_matrix_[x][y] += 1

        self.TP = self.confusion_matrix_[0][0]
        self.FP = self.confusion_matrix_[0][1]
        self.FN = self.confusion_matrix_[1][0]
        self.TN = self.confusion_matrix_[1][1]

    def confusion_matrix(self):
        return self.confusion_matrix_
    
    def accuracy(self):
        print(f'accuracy:\n{round((self.TP + self.TN) / self.confusion_matrix_.sum(),2)}\n')
    
    def precision(self):
        print('precision:')
        self.precision_0 = self.TP / (self.TP + self.FN)
        self.precision_1 = self.TN / (self.TN + self.FP)
        print(f'0: {round(self.precision_0,2)}\n1: {round(self.precision_1,2)}\n')
    
    def recall(self):
        print('recall:')
        self.recall_0 = self.TP / (self.TP + self.FP)
        self.recall_1 = self.TN / (self.TN + self.FN)
        print(f'0: {round(self.recall_0,2)}\n1: {round(self.recall_1,2)}\n')

    def f1(self):
        print('f1-score:')
        f1_0 = 2 / (1 / self.precision_0 + 1 / self.recall_0)
        f1_1 = 2 / (1 / self.precision_1 + 1 / self.recall_1)
        print(f'0: {round(f1_0,2)}\n1: {round(f1_1,2)}\n')

In [95]:
my = MyMetrics(y_test, predict)
my.precision()
my.recall()
my.f1()
my.accuracy()

precision:
0: 0.58
1: 0.5

recall:
0: 0.02
1: 0.99

f1-score:
0: 0.04
1: 0.67

accuracy:
0.5



In [96]:
print(classification_report(y_test,predict))

              precision    recall  f1-score   support

           0       0.58      0.02      0.04      8675
           1       0.50      0.99      0.67      8704

    accuracy                           0.50     17379
   macro avg       0.54      0.50      0.35     17379
weighted avg       0.54      0.50      0.35     17379



# Задание 3.2

- Самостоятельно реализуйте kNN.

Сделано в Lab4_MyKNN.ipynb

# Задание 4*

- Самостоятельно реализуйте SVM с возможностью выбирать ядро.